invokeai/app/services/config/config_default.py

```diff
@@ -428,33 +440,33 @@ def load_and_migrate_config(config_path: Path) -> InvokeAIAppConfig:
     """
     assert config_path.suffix == ".yaml"
     with open(config_path, "rt", encoding=locale.getpreferredencoding()) as file:
-        loaded_config_dict = yaml.safe_load(file)
+        current_config_dict = yaml.safe_load(file)
 
-    assert isinstance(loaded_config_dict, dict)
+    assert isinstance(current_config_dict, dict)
+    new_config_dict = copy.deepcopy(current_config_dict)
 
-    if "InvokeAI" in loaded_config_dict:
-        # This is a v3 config file, attempt to migrate it
+    if "InvokeAI" in current_config_dict:
+        new_config_dict = migrate_v3_config_dict(new_config_dict)  # pyright: ignore [reportUnknownArgumentType]
+
+    if new_config_dict["schema_version"] == "4.0.0":
+        new_config_dict = migrate_v4_0_0_config_dict(new_config_dict)
+
+    if new_config_dict["schema_version"] == "4.0.1":
+        new_config_dict = migrate_v4_0_1_config_dict(new_config_dict)
+
+    config_dict_updated = new_config_dict != current_config_dict
+    if config_dict_updated:
         shutil.copy(config_path, config_path.with_suffix(".yaml.bak"))
-        try:
-            # loaded_config_dict could be the wrong shape, but we will catch all exceptions below
-            migrated_config = migrate_v3_config_dict(loaded_config_dict)  # pyright: ignore [reportUnknownArgumentType]
-        except Exception as e:
-            shutil.copy(config_path.with_suffix(".yaml.bak"), config_path)
-            raise RuntimeError(f"Failed to load and migrate v3 config file {config_path}: {e}") from e
-        migrated_config.write_file(config_path)
-        return migrated_config
-
-    if loaded_config_dict["schema_version"] == "4.0.0":
-        loaded_config_dict = migrate_v4_0_0_config_dict(loaded_config_dict)
-        loaded_config_dict.write_file(config_path)
 
     # Attempt to load as a v4 config file
     try:
         # Meta is not included in the model fields, so we need to validate it separately
-        config = InvokeAIAppConfig.model_validate(loaded_config_dict)
+        config = InvokeAIAppConfig.model_validate(new_config_dict)
         assert (
             config.schema_version == CONFIG_SCHEMA_VERSION
         ), f"Invalid schema version, expected {CONFIG_SCHEMA_VERSION}: {config.schema_version}"
+        if config_dict_updated:
+            config.write_file(config_path)
         return config
     except Exception as e:
         raise RuntimeError(f"Failed to load config file {config_path}: {e}") from e
```

Comment:
We should call `model_validate()` _before_ moving the old config file.

Also, the block of code wrapped by the try-except right now feels sort of random

Reply:
IIRC, the try-catch block is intended to present a more user-friendly error. If it's not wrapped, there's a pydantic validation error on startup, which is confusing. With the wrap, there's a good plain-English error: "Failed to load config file...".